core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1413,55 +1412,108 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
+              F.race(
+                F.sleep(timeout),
+                supply.acquireN(n)
+              ).flatMap {
                 case Left(_) =>
-                  if (acc.nonEmpty)
-                    Pull.output1(acc) >> resetTimerAndGo(Chunk.empty)
-                  else
-                    go(Chunk.empty, next, hasTimedOut = true)
-                case Right(c) if hasTimedOut =>
-                  // it has timed out without reset, so acc is empty
-                  val (toEmit, rest) =
-                    if (c.size < n) Pull.output1(c) -> Chunk.empty
-                    else resize(c, Pull.done)
-                  toEmit >> resetTimerAndGo(rest)
-                case Right(c) =>
-                  val newAcc = acc ++ c
-                  if (newAcc.size < n)
-                    go(newAcc, next)
-                  else {
-                    val (toEmit, rest) = resize(newAcc, Pull.done)
-                    toEmit >> resetTimerAndGo(rest)
+                  supply.acquire.flatMap { _ =>
+                    supply.available.flatMap { m =>
+                      val k = m.min(n - 1)
+                      supply.tryAcquireN(k).map {
+                        case true  => k + 1
+                        case false => 1
+                      }
+                    }
                   }
+                case Right(_) =>
+                  F.pure(n)
               }
-          }
 
-        timedPull.timeout(timeout) >> go(Chunk.empty, timedPull)
+            def dequeueN(n: Int): F2[Option[Vector[O]]] =
+              acquireSupplyUpToNWithin(n.toLong, timeout).flatMap { n =>
+                buffer
+                  .modify { buf =>
+                    if (buf.data.size >= n) {
+                      val (head, tail) = buf.data.splitAt(n.toInt)
+                      (buf.copy(tail), buf.copy(head))
+                    } else {
+                      (buf.copy(Vector.empty), buf)
+                    }
+                  }
+                  .flatMap { buf =>
+                    demand.releaseN(buf.data.size.toLong).flatMap { _ =>
+                      buf.endOfSupply match {
+                        case Some(Left(error)) =>
+                          F.raiseError(error)
+                        case Some(Right(_)) if buf.data.isEmpty =>
+                          F.pure(None)
+                        case _ =>
+                          F.pure(Some(buf.data))
+                      }
+                    }
+                  }
+              }
+
+            def endSupply(result: Either[Throwable, Unit]): F2[Unit] =
+              buffer.update(_.copy(endOfSupply = Some(result))) *> supply.releaseN(Int.MaxValue)
+
+            def endDemand(result: Either[Throwable, Unit]): F2[Unit] =
+              buffer.update(_.copy(endOfDemand = Some(result))) *> demand.releaseN(Int.MaxValue)
+
+            val enqueueAsync = F.start {
+              this
+                .evalMap(enqueue)
+                .takeWhile(_ == true)
```

Comment:
```suggestion
                .forall(_)
```
Because `forall` is a fold, it discards elements (chunks of booleans) sooner.

Reply:
:+1: